### PR TITLE
[Bug] [Workflow] Fix workflow failure strategy bug

### LINKED ISSUE #11781: [Bug] [Workflow] Workflow failure strategy does not work as expected
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

* Failure strategy `continue` does not work for serial tasks. Although the doc doesn't say it could work for serial tasks, it should work for it instead of only for parallel tasks.
* In many scenarios, we need a `clean-up` task at the end of a workflow. Without failure strategy working as expected, we could never achieve this.
* This bug exists in both `3.0.0-release` and `dev`.
image
image

### What you expected to happen

* Failure strategy `continue` is supposed to work for all kinds of workflows.
* We need to fix the bug and change the docs.

### How to reproduce

* Already described above.

### Version

dev

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by EricGao888**
## Purpose of the pull request

* This PR closes: #11781 

## Brief change log

* Make `Failure Strategy` `continue` work.

* For details, see https://github.com/apache/dolphinscheduler/issues/11781

## Verify this pull request

* Verified by unit tests and manual tests.

### CODE REVIEW

**Review by SbloodyS — DISMISSED** [review]
LGTM

**Review by stalary — DISMISSED** [review]
LGTM

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java [hunk 1]
```diff
@@ -1808,7 +1812,8 @@ public void submitStandByTask() throws StateEventHandleException {
                 getPreVarPool(task, preTask);
             }
             DependResult dependResult = getDependResultForTask(task);
-            if (DependResult.SUCCESS == dependResult) {
+            if (DependResult.SUCCESS == dependResult
+                    || processInstance.getFailureStrategy() == FailureStrategy.CONTINUE) {
```
Comment (ruanwenjun) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java]:
```suggestion
            if (DependResult.SUCCESS == dependResult || (DependResult.FAILED == dependResult && processInstance.getFailureStrategy() == FailureStrategy.CONTINUE)) {
```

**Review by ruanwenjun — CHANGES_REQUESTED** [review]

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java [hunk 2]
```diff
@@ -1808,7 +1812,8 @@ public void submitStandByTask() throws StateEventHandleException {
                 getPreVarPool(task, preTask);
             }
             DependResult dependResult = getDependResultForTask(task);
-            if (DependResult.SUCCESS == dependResult) {
+            if (DependResult.SUCCESS == dependResult || ((DependResult.WAITING != dependResult)
```
Comment (EricGao888) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java:1815]: What about changing it to `if ((DependResult.WAITING != dependResult)
                    && (processInstance.getFailureStrategy() == FailureStrategy.CONTINUE))` ? @ruanwenjun

**Comment by caishunfeng** [comment]
I found the UT case `WorkflowExecuteRunnableTest.testCheckSerialProcess`, can you add a UT case for it?

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java (hunk 2 above)
Comment (ruanwenjun) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java]: Yes, it's ok.

**Comment by EricGao888** [comment]
> I found the UT case `WorkflowExecuteRunnableTest.testCheckSerialProcess`, can you add a UT case for it?

@caishunfeng Of course. I will take a look. Thanks for the reminder : )

**Review by ruanwenjun — DISMISSED** [review]
LGTM, it will be better if you can add UT.

**Comment by EricGao888** [comment]
> LGTM, it will be better if you can add UT.

Sure, I'm working on it.

**Comment by EricGao888** [comment]
Refactored some code for better readability and testability. Related unit tests added : ) PTAL @ruanwenjun @caishunfeng @SbloodyS @stalary

**Review by SbloodyS — APPROVED** [review]
+1

**Review by caishunfeng — APPROVED** [review]
LGTM :+1:

**Review by lenboo — CHANGES_REQUESTED** [review]
According to the docs about this feature, I think there are disagreements.

**Comment by EricGao888** [comment]
> According to the docs about this feature, I think there are disagreements.

@lenboo May I ask whether the `continue` strategy was designed only for `parallel` execution, or there is an incorrect description in docs? If the latter, I will fix the docs. If the former, we need further discussion. However, IMHO, I think `continue` strategy should not only work for `parallel`. Sometimes we need `clean up tasks`. As described in issue.

image

**Comment by lenboo** [comment]
> > According to the docs about this feature, I think there are disagreements.
> 
> @lenboo May I ask whether the `continue` strategy was designed only for `parallel` execution, or there is an incorrect description in docs? If the latter, I will fix the docs. If the former, we need further discussion. However, IMHO, I think `continue` strategy should not only work for `parallel`. Sometimes we need `clean up tasks`. As described in issue.
> 
> image

 - Yes, the continue strategy was designed only for parallel tasks.
 - I can understand the scenario you mentioned. Indeed, we need solution to this scenario in another discussion.

**Comment by EricGao888** [comment]
> > > According to the docs about this feature, I think there are disagreements.
> > 
> > 
> > @lenboo May I ask whether the `continue` strategy was designed only for `parallel` execution, or there is an incorrect description in docs? If the latter, I will fix the docs. If the former, we need further discussion. However, IMHO, I think `continue` strategy should not only work for `parallel`. Sometimes we need `clean up tasks`. As described in issue.
> > image
> 
> * Yes, the continue strategy was designed only for parallel tasks.
> * I can understand the scenario you mentioned. Indeed, we need solution to this scenario in another discussion.

@lenboo Thanks for the reply. I've add the topic to community bi-weekly conference topic list, looking forward to the discussion tonight : )

**Comment by EricGao888** [comment]
I'm closing it temporarily. Will submit a design later.

**Pull request closed without merging**